base/experimental.jl

```diff
@@ -256,4 +257,35 @@ end
 # OpaqueClosure
 include("opaque_closure.jl")
 
+"""
+    Experimental.@overlay mt [function def]
+
+Define a method and add it to the method table `mt` instead of to the global method table.
+This can be used to implement a method override mechanism. Regular compilation will not
+consider these methods, and you should customize the compilation flow to look in these
+method tables (e.g., using [`Core.Compiler.OverlayMethodTable`](@ref)).
+
+"""
+macro overlay(mt, def)
+    def = macroexpand(__module__, def) # to expand @inline, @generated, etc
+    if !isexpr(def, [:function, :(=)]) || !isexpr(def.args[1], :call)
+        error("@overlay requires a function Expr")
+    end
+    def.args[1].args[1] = Expr(:overlay, mt, def.args[1].args[1])
+    esc(def)
+end
+
+"""
+    Experimental.@MethodTable(name)
+
+Create a new MethodTable in the current module, bound to `name`. This method table can be
+used with the [`Experimental.@overlay`](@ref) macro to define methods for a function without
+adding them to the global method table.
+"""
+macro MethodTable(name::Symbol)
+    esc(quote
+        const $name = ccall(:jl_new_method_table, Any, (Any, Any), $(quot(name)), $(__module__))
```

Comment:
ccall is somewhat harder on the compiler as a return value, since we can't optimize it but must compile it. Lifting it into a function call should avoid those problems.
```suggestion
    new_mt() = ccall(:jl_new_method_table, Any, (Any, Any), name, __module__)
    return esc(:(const $name = $new_mt()))
```